### PR TITLE
data loader

### PR DESCRIPTION
**Pull request by lizli502**
Finished functionality of data loader

### CODE REVIEW

File: src/data_loader.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import PIL
+import torch
+import random
+import collections
+import numpy as np
+import pandas as pd
+
+
+
+class ImageSegmentaionDataset():
```
Comment (grvk) [inline src/data_loader.py]: According to this [tutorial](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html), it's customary to extend one of the Dataset classes from pytorch.
For example:
```
class FaceLandmarksDataset(Dataset):
    """Face Landmarks dataset."""
```
Could you extend torch.utils.data.Dataset?

File: src/data_loader.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    """Image Segmentation dataset"""
+    
+    def __init__(self, image_dict, loader_type, replacement=True, transform=None):
+        """
+        Args:
+        image_dict (dict): a dict of the following format:
+                        {"ships": {"imageId": "segmentation pixels"}， “w/ships”: ["imageId"]}
+        loader_type: (string)"classification" or "segmentation"
```
Comment (grvk) [inline src/data_loader.py]: Could you split up this dataset into two?
For example, ImageSegmentationDataset and ImageClassificationDataset. Right now the naming "ImageSegmentaionDataset" suggests only one type, so presence of "loader_type" is confusing.

It might make sense to have a base class with all common functionality and extend it to ImageSegmentation and ImageClassification ones, which will override type-specific methods only.

File: src/data_loader.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        replacement: (bool) sample with/without replacement
+        transform (callable, optional): Optional transform to be applied on a sample.
+        """
+        self.image_dict = image_dict
```
Comment (grvk) [inline src/data_loader.py:31]: Could we use a bit more descriptive naming? Also, could we not use "dict" since it's an [anti-pattern](https://www.codewithjason.com/variable-name-anti-patterns/)? (see "Naming a Variable after Its Type"

File: src/data_loader.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        self.loader_type = loader_type
+        self.replacement = replacement
+        # images_list is the name 
+        self.images_list = list(image_dict["ships"].keys()) + image_dict["w/ships"]
```
Comment (grvk) [inline src/data_loader.py]: Same comment as above

File: src/data_loader.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        random.shuffle(self.images_list) 
+        self.transform = transform
+    
+    def __len__(self):
+        return len(self.images_list)
+    
+    def rle_decode(self, mask_rle, shape=(768, 768)):
```
Comment (grvk) [inline src/data_loader.py]: It makes sense to take this function out to a separate class, say: RLEHandler. The reason for it is because we could possibly use this function beyond the scope of this class. Additionally, (later) we'll need to implement rle_encode to prepare submission file. So all this functionality should be in one place

File: src/data_loader.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        img = PIL.Image.open(image_path)
+        image_mat = np.array(img)
+        return image_mat
+        
+    def __getitem__(self, idx):
+        image_id = self.images_list[idx]
+        no_ship_list = image_dict["w/ships"]
+        image_mat = self.retrive_image_matrix(image_id)
+        # dataloader for classification
+        if self.loader_type == "classification":
```
Comment (grvk) [inline src/data_loader.py:80]: The output for classification should an array of length 2.
[1, 0] means it belongs to the first category (no ships)
[0, 1] means it belongs to the second category (ships present)

It also needs to be a tuple:
i.e. (<768x768 matrix of 0-255 values>, [0,1])

File: src/data_loader.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            if image_id in no_ship_list:
+                sample = {"image": image_mat, "class": 0}
+            else:
+                sample = {"image": image_mat, "class": 1}
+        
+        # dataloader for segmentation 
+        else: 
+            mask_list = image_dict["ships"][image_id]
```
Comment (grvk) [inline src/data_loader.py]: If I understand correctly, this code assumes that we always return a mask for the image from the list of images with ships. This is not the case.
Example of outputs:
tuples*

without ships:
(<768x768 matrix of 0-255 values>, <768x768 matrix of 0's>)

with ships:
(<768x768 matrix of 0-255 values>, <768x768 matrix of 0's and 1's>)

**Review by grvk — CHANGES_REQUESTED** [review]

File: src/data_loader.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        all_masks = np.zeros((768, 768), dtype = np.int16)
+        #if isinstance(in_mask_list, list):
+        for mask in in_mask_list:
+            if isinstance(mask, str):
+                all_masks += self.rle_decode(mask)
+        return np.expand_dims(all_masks, -1)
+    
+    
+    def retrive_image_matrix(self, image_id):
+        # google platform (TODO!!!!!!!!!)
```
Comment (grvk) [inline src/data_loader.py]: Could you clarify the comment? What should be done?

File: src/data_loader.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+import collections
+import numpy as np
+import pandas as pd
+
+
+
+class ImageSegmentaionDataset():
+    """Image Segmentation dataset"""
+    
+    def __init__(self, image_dict, loader_type, replacement=True, transform=None):
```
Comment (grvk) [inline src/data_loader.py]: Could we add one more parameter to accept path to images?

File: src/data_loader.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        #if isinstance(in_mask_list, list):
+        for mask in in_mask_list:
+            if isinstance(mask, str):
+                all_masks += self.rle_decode(mask)
+        return np.expand_dims(all_masks, -1)
+    
+    
+    def retrive_image_matrix(self, image_id):
+        # google platform (TODO!!!!!!!!!)
+        image_path = path.join(path.dirname(__file__), '../data/train_v2/', image_id)
```
Comment (grvk) [inline src/data_loader.py]: Per the comment before, we should not hardcode and use a path passed to the constructor, so this line should be like:
```
path.join(self.path_to_images, image_id)
```

File: src/data_loader.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+        # kaggle image path
+        #image_path = "/kaggle/input/airbus-ship-detection/train_v2/" + image_id
```
Comment (grvk) [inline src/data_loader.py]: I would take this out. If we need to, we can add it to Readme? 
This class should be environment-agnostic.

File: src/data_loader.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        replacement: (bool) sample with/without replacement
+        transform (callable, optional): Optional transform to be applied on a sample.
+        """
+        self.image_dict = image_dict
+        self.loader_type = loader_type
+        self.replacement = replacement
+        # images_list is the name 
+        self.images_list = list(image_dict["ships"].keys()) + image_dict["w/ships"]
+        random.shuffle(self.images_list) 
+        self.transform = transform
```
Comment (grvk) [inline src/data_loader.py:36]: Our segmentation dataset cannot work with ```torchvision.transforms.Resize``` out of the box since besides scaling an image we also need to scale the mask.

Can you create a separate field to store this transform, i.e. ```self.resize_transform```. By default, it can be ```None```. And if passed resize transform, we add this step both to the mask(make sure it works with our output format, since the expected input is PIL image) and the image

File: src/data_loader.py (hunk 7 above)
Comment (grvk) [reply src/data_loader.py]: See comment above about ```resize``` transform. In this case, the size of both mask and input image need to be scaled and equal

Comment (grvk) [reply src/data_loader.py]: More ideas. The resize transform will use bilinear upscale, so the resulting values might be between 0 and 1. To keep mask values 0 or 1, filter the output with the following threshold:
if pixel value >= 0.5 > make it 1
if pixel value < 0.5 make it 0

File: src/data_loader.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+            mask = self.masks_as_image(mask_list) 
+            sample = {"image": image_mat, "mask": mask}
+
+        
+        # remove image id if sample without replacement
+        if self.replacement:
+            self.images_list.remove(image_id)
+
+        if self.transform:
+            sample = self.transform(sample)
```
Comment (grvk) [inline src/data_loader.py:103]: Don't forget to update resize transforms :)

File: src/data_loader.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        random_idx = random.randint(0,n_image_list)
+        sample = self.__getitem__(random_idx)
+        return sample
+
+
+
+# helper functions
+
+# input: file name of the segmentation csv file
+# output: a dictionary {"ships": {"imageId": list of segmentation pixels}， “w/ships”: ["imageId"]}
```
Comment (grvk) [inline src/data_loader.py]: We don't really need list. We will always have only one string, so we can either:
- Remove list and have ```{"ships": {"image_id": "rle_encoding"}, "w/ships":["image_id"]}```
- Or ```{"ships": [{"image_id":"rle_encoding"}], "w/ships":["image_id"]}```. Originally, the thought behind this was to keep same format for ships and w/ships, as well as simplify working with lists (i.e. we don't have to use keys() anywhere etc.).

Probably, at this point it makes sense to go with 1st way and remove unnecessary list. We seem to be working with keys() and random sampling, splitting just fine, so option #2 might be more comlicated

File: src/data_loader.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+    return(image_dict)
+
+
+# preprocess the image dict to have same images with/without ships
+def process_images_dict(image_dict):
+    n = len(image_dict["ships"])
+    image_dict["w/ships"] = random.choices(image_dict["w/ships"], k=n)
+    return image_dict
+
+
```
Comment (grvk) [inline src/data_loader.py]: The code below needs to be removed. Otherwise, it's run every time we try to import the file.
If we want to add some testing, see an example [here](https://github.com/grvk/airbus-challenge/blob/dev/george/trainer/tests/tests_trainer.py). Since we're short on time though, as long as we test it manually, automated tests can be skipped

File: src/data_loader.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        n_image_list = len(self.images_list)
+        
+        # randomly choose a idx
+        random_idx = random.randint(0,n_image_list)
+        sample = self.__getitem__(random_idx)
+        return sample
+
+
+
+# helper functions
```
Comment (grvk) [inline src/data_loader.py]: Let's create a file called something like "helper.py" (if you can come up with better name, I'm down :D ) and move these functions there. They shouldn't be in dataloader file really

**Review by grvk — CHANGES_REQUESTED** [review]

File: src/data_loader.py (hunk 1 above)
Comment (lizli502) [reply src/data_loader.py]: Done

File: src/data_loader.py (hunk 2 above)
Comment (lizli502) [reply src/data_loader.py]: Sorry for the confusion..... I like your suggestion, but in order to save time, I changed the name of the class to be AirbusDataset and still include the "loader_type".

File: src/data_loader.py (hunk 3 above)
Comment (lizli502) [reply src/data_loader.py:31]: I think I will change that later if I had time because I cannot think of another appropriate name now......

File: src/data_loader.py (hunk 4 above)
Comment (lizli502) [reply src/data_loader.py]: I changed "images_list" to be "image_filenames".

File: src/data_loader.py (hunk 5 above)
Comment (lizli502) [reply src/data_loader.py]: Done, write the RLEHandler class in the file rle_handler.py

File: src/data_loader.py (hunk 6 above)
Comment (lizli502) [reply src/data_loader.py:80]: DONE

File: src/data_loader.py (hunk 8 above)
Comment (lizli502) [reply src/data_loader.py]: I am not sure where is the path for all the images, so I want you to modify it to get the path for the image folder.

File: src/data_loader.py (hunk 7 above)
Comment (lizli502) [reply src/data_loader.py]: 1. This is indeed a bug and I've fixed it.
2.  mask: (1, 768, 768)       input: (3, 768, 768) 
3*.  As far as I concern, it is unlikely for the resulting values to be between 0 and 1, so I ignored it for now.

File: src/data_loader.py (hunk 9 above)
Comment (lizli502) [reply src/data_loader.py]: DONE!

File: src/data_loader.py (hunk 10 above)
Comment (lizli502) [reply src/data_loader.py]: Done

File: src/data_loader.py (hunk 11 above)
Comment (lizli502) [reply src/data_loader.py]: Done

File: src/data_loader.py (hunk 12 above)
Comment (lizli502) [reply src/data_loader.py:36]: * I am lost on this part
I just add a line in the constructor ```self.resize_transform = None``` for now.

File: src/data_loader.py (hunk 14 above)
Comment (lizli502) [reply src/data_loader.py]: I think we do need the list because every element in the list is a rle_encoding for one ship mask.  For the later decoding, we do that one mask at a time.
